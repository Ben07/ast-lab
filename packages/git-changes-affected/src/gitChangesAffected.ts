import resolver from 'enhanced-resolve';
import huntAffected from 'hunt-affected';
import { ParserOptions } from '@babel/parser';
import _debug from 'debug';
import { getGitDiffs } from './getGitDiffs';
import IncludesFilePlugin from './includesFilePlugin';
import { GIT_OPERATION } from './constants';
import { Change } from './types';
import getRevisionFile from './getRevisionFile';
import denodeify from './denodeify';
import getTrackedFiles from './getTrackedFiles';
import hasExt from './hasExt';
import getAbsolutePath from './getAbsolutePath';
import getChangedEntries from './getChangedEntries';

const debug = _debug('git-changes-affected:affected');
const DEFAULT_EXTENSIONS = ['.js', '.jsx', '.ts', '.tsx'];

function changesAffected(commit: string, changes: Change[], { alias, modules, parserOptions, paths }: Options) {
  const extensions = DEFAULT_EXTENSIONS;
  const trackedFiles = getTrackedFiles(commit, paths).filter(file => hasExt(file, extensions)).map(getAbsolutePath);
  const entries = getChangedEntries(changes, parserOptions);
  debug(`${commit} entries: ${JSON.stringify(entries)}`);
  return huntAffected(
    trackedFiles,
    entries,
    {
      loader: (file: string) => Promise.resolve(getRevisionFile(commit, file)),
      resolver: denodeify(resolver.create({
        extensions,
        alias,
        modules,
        plugins: [new IncludesFilePlugin(trackedFiles, extensions)]
      })),
      parserOptions
    }
  );
}

type Options = {
  modules?: string[],
  // extensions?: string[],
  alias?: { [alias: string]: string },
  parserOptions?: ParserOptions,
  paths?: string[]
};
export default async function gitChangesAffected(commit: string, opts: Options = {}) {
  const extensions = DEFAULT_EXTENSIONS;
  const diffs = getGitDiffs(commit);
  const befores = [] as Change[];
  const afters = [] as Change[];
  diffs.forEach(({ source, target, operation }) => {
    if (
      operation !== GIT_OPERATION.new &&
      hasExt(source.file, extensions)
    ) {
      befores.push(source);
    }
    if (
      operation !== GIT_OPERATION.delete && 
      hasExt(target.file, extensions)
    ) {
      afters.push(target);
    }
  });
  const affected = await changesAffected(`${commit}~1`, befores, opts);
  debug('before affected:', affected);
  const toMerge = await changesAffected(commit, afters, opts);
  debug('after affected:', toMerge);
  Object.keys(toMerge).forEach(mod => {
    const members = toMerge[mod];
    affected[mod] =affected[mod] ? new Set([
      ...Array.from(affected[mod]),
      ...Array.from(members)
    ]) : members;
  });
  return affected;
}