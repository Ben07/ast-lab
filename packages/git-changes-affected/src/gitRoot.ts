import exec from './exec';

export default function gitRoot() {
  return exec('git rev-parse --show-toplevel');
}