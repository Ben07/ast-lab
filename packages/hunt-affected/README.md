hunt-affected
===

Detect where your file exports are used and potentially afffected by it's changes.

## Quick start

Intalling via npm:
```sh
npm i hunt-affected
```

And feed the function with a list of absolute file paths you would like to check, and the entry points like this:
```javascript
const huntAffected = require('hunt-affected');

huntAffected(
  ['a.js', 'b.js', 'c.js'],
  [{ source: 'a.js', name: 'default' }]
);
```

To support things like `JSX`, `flow`, `dynamic imports`, etc.. You will need to enable `@babel/parser` plugins:
```javascript
huntAffected(
  ['a.js', 'b.js', 'c.js'],
  [{ source: 'a.js', name: 'default' }]
  {
    parserOptions: {
      plugins: ['jsx', 'dynamicImport']
    }
  }
);
```

All the options in `parserOptions` will be passed to `@babel/parser` directly. `@babel/parser` options can be found [here](https://babeljs.io/docs/en/babel-parser).

By default, it will try to read file with NodeJs default file system and decode them with `utf-8`.

You may replace this behavior by passing a customised `loader` function:
```javascript
huntAffected(
  ['a.js', 'b.js', 'c.js'],
  [{ source: 'a.js', name: 'default' }]
  {
    loader: async (path) {
      return await myWayToReadFile(path);
    }
  }
);
```

And when it tries to resolve file imported module paths to absolute file path, it will use Webpack's [enhanced-resolve](http://github.com/webpack/enhanced-resolve) by default, and tries to resolve to real files.

You may replace this behavior by passing a customised `resolver` function:
```javascript
huntAffected(
  ['a.js', 'b.js', 'c.js'],
  [{ source: 'a.js', name: 'default' }]
  {
    resolver: async (base: string, target: string) => {
      return 'resolved/file/path.js';
    }
  }
);
```