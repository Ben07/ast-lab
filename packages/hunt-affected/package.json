{
  "name": "hunt-affected",
  "version": "1.0.0-alpha",
  "description": "",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "../../node_modules/.bin/tsc --build --verbose ./",
    "test": "../../node_modules/.bin/jest  --config=./jest.config.js"
  },
  "keywords": [
    "affected",
    "dependency",
    "change"
  ],
  "files": [
    "./dist"
  ],
  "author": "Jennie Ji <jennie.ji@hotmail.com>",
  "repository": {
    "url": "https://github.com/JennieJi/ast-lab",
    "type": "git"
  },
  "bugs": {
    "url": "https://github.com/JennieJi/ast-lab/issues"
  },
  "license": "MIT",
  "devDependencies": {
    "@babel/parser": "^7.7.5",
    "@types/debug": "^4.1.5",
    "@types/enhanced-resolve": "^3.0.6",
    "ast-lab-types": "1.0.0-alpha"
  },
  "dependencies": {
    "debug": "^4.1.1",
    "enhanced-resolve": "^4.1.1",
    "es-stats": "1.0.0-alpha"
  }
}
