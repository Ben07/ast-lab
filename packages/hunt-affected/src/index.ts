import huntAffected from './huntAffected';

export { default as fileDepMap } from './fileDepMap';
export { default as mergeDepMap } from './mergeDepMap';
export { default as visitDepMap, Visited as Affected } from './visitDepMap';
export default huntAffected;