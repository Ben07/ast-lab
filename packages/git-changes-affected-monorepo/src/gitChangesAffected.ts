import { Affected } from 'hunt-affected';
import {
  huntRevisionImpact,
  getDiffEntries,
  ParserOptions,
} from 'git-changes-affected';

function mergeAffected(base: Affected, toMerge: Affected): void {
  Object.keys(toMerge).forEach(mod => {
    const members = toMerge[mod];
    base[mod] = base[mod]
      ? new Set([...Array.from(base[mod]), ...Array.from(members)])
      : members;
  });
}

export type Revisions = {
  /** The end result revision. Default to `HEAD`. */
  to?: string;
  /** The revision to compare from. Default to one commit before `to` revision. */
  from?: string;
};

export default async function gitChangesAffected(
  resolves: Array<{
    /** Limit paths of tracked files to check with. By default it will check all the git tracked files. */
    paths: string[];
    /** Paths where to look for JS modules, if you have customised modules other than npm's `node_modules`. */
    modules?: string[];
    // extensions?: string[],
    /** Module alias to a path */
    alias?: { [alias: string]: string };
    /** `@babel/parser` options for parsing file to AST */
    parserOptions?: ParserOptions;
    /** Filter file extensions, eg. `['.js', '.jsx', '.ts', '.tsx]`. */
    extensions?: string[];
  }>,
  opts: Revisions = {}
): ReturnType<typeof huntRevisionImpact> {
  const to = opts.to || `HEAD`;
  const from = opts.from || `${to}~1`;
  const { befores, afters } = getDiffEntries(opts);
  const affected = {} as Affected;
  await Promise.all(
    resolves.map(async resolveOptions => {
      mergeAffected(
        affected,
        await huntRevisionImpact(from, befores, resolveOptions)
      );
      mergeAffected(
        affected,
        await huntRevisionImpact(to, afters, resolveOptions)
      );
    })
  );
  return affected;
}
