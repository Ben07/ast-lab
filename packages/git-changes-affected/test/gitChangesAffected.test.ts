import path from 'path';
import gitRoot from '../src/gitRoot';
import gitChangesAffected from '../src/gitChangesAffected';
import { Affected } from 'hunt-affected';


const ROOT = gitRoot();
function relativeEntries(affected: Affected) {
  const updated = {} as Affected;
  Object.keys(affected).forEach(p => {
    updated[path.relative(ROOT, p)] = affected[p];
  });
  return updated;
}

describe('gitChangesAffected', () => {
  test('can work', async () => {
    const case1 = await gitChangesAffected('e42f4f6a08569afd48540759cc604aab2b8c02f3', {
      parserOptions: {
        plugins: ['typescript']
      },
      paths: [
        'packages/es-stats/src'
      ]
    });
    expect(relativeEntries(case1)).toMatchSnapshot();

    const case2 = await gitChangesAffected('ec7b4c814c17c4b9a9467ba0c876983a9b3cd6ec', {
      parserOptions: {
        plugins: ['typescript']
      }
    });
    expect(relativeEntries(case2)).toMatchSnapshot();
  });
});