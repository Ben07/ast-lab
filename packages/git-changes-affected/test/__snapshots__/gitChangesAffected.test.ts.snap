// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gitChangesAffected can work 1`] = `
Object {
  "packages/es-stats/src/index.ts": Set {},
  "packages/es-stats/src/visitors/exports.ts": Set {},
}
`;

exports[`gitChangesAffected can work 2`] = `
Object {
  "packages/es-stats/src/extractStats.ts": Set {
    "default",
  },
  "packages/es-stats/src/fileStats.ts": Set {
    "default",
  },
  "packages/es-stats/src/index.ts": Set {},
  "packages/es-stats/src/visitors/exports.ts": Set {
    "default",
  },
  "packages/es-stats/src/visitors/imports.ts": Set {
    "default",
  },
  "packages/es-stats/src/visitors/rootRelation.ts": Set {
    "default",
  },
  "packages/hunt-affected/src/fileDepMap.ts": Set {
    "default",
  },
  "packages/hunt-affected/src/huntAffected.ts": Set {
    "default",
  },
  "packages/hunt-affected/src/index.ts": Set {},
  "packages/hunt-affected/src/mergeDepMap.ts": Set {
    "default",
  },
  "packages/hunt-affected/src/visitDepMap.ts": Set {
    "default",
  },
}
`;
